### ISSUE TITLE
Colcon logging incurs significant overhead on Windows

### ISSUE THREAD

**Issue by KazNX** [colcon/colcon-core#211]
Hi,

I've noted that colcon incurs significant overheads when logging on Windows. This is on top of it being significantly slower on Windows vs Linux.

Attached is a helloworld colcon project which deliberately generates log spam during CMake configuration. The `README.md` file contains instructions on how to run colcon to show differences between default logging, console only logging and no logging as well as the performance summary, duplicated in the table below.

OS        | File logging  | Console logging | No logging
--------- | ------------- | --------------- | --------------
Windows   | 35.2s         | 12.8s           | 12.2s
Linux     | 0.55s         | 0.44s           | 0.46s

And yes, Linux is generally nearly two orders of magnitude faster.

[slowcolcon.zip](https://github.com/colcon/colcon-core/files/3523303/slowcolcon.zip)

**Comment by KazNX** [comment]
Addendum: I've timed just executing cmake normally to generate the log spam and this performs reasonably at ~1.2s. Still worse than Linux, but at least not by several orders of magnitude. This suggests that the drastic difference in timing between Windows an Linux appears to all be incurred by colcon.

To replicate this result:
1. Run the build process as described in the README.md of slowcolcon.zip
2. Time the execution of `cmake build\slow`

I used the following PowerShell command for this: `Measure-Command { cmake build\slow }`

**Comment by dirk-thomas** [comment]
Thanks for reporting this.

I think Windows is in general significantly slower for IO operation than Linux / macOS. That being said it should certainly not be *that* bad.

There are two locations where `colcon-output` writes the captured output to files:
* https://github.com/colcon/colcon-output/blob/3a94106c28d64f66aaad38653dae593312b358a8/colcon_output/event_handler/event_log.py#L42-L46
* https://github.com/colcon/colcon-output/blob/3a94106c28d64f66aaad38653dae593312b358a8/colcon_output/event_handler/log.py#L96-L97

The second one you already disabled when passing `--event-handlers log-`. @KazNX Can you please also try to measure the time when you additionally disable the `event_log` by passing `--event-handlers log- event_log-`.

If that gets the remaining time of ~12s much further down I would suggest we try to keep the file handles open for file which are repeatedly being written to. That will likely make this faster.

**Comment by dirk-thomas** [comment]
It would be great if you could also compare the performance of this branch (https://github.com/colcon/colcon-output/tree/dirk-thomas/keep-file-handles-open) which keeps the all file handles open between writes. That should speed up the `log` event handler. If that does help we can apply the same logic to the `event_log` event handler.

**Comment by KazNX** [comment]
I agree, Linux IO is much better than Windows, but shouldn't be by that much. I will try the linked changelist later. I've provided additional timing information below. The first two are without the changes in the branch keep-file-handles-open. The last one is with keep-file-handles-open, and no `--event-handlers` argument.

OS | `--event-handlers log- event_log-` | `--event-handlers log- event_log- console_direct+` | keep-file-handlers-open
--- | -------------------------------------- | ------------------------------------------------------- | ----
Windows | 6.2s | 5.8s | 1.9s*

\* There is a pause after build timing is reported, but before python exists which is not measured by this time. This pause appears to be present in the other runs too, so your branch is a clear winner. Nice work and thanks for the speedy response.

I suspect the remaining overheads are due to MSBuild not colcon. I have noticed that on Windows there seems to be redundant work being done for a clean target. Specifically, building a clean target on Windows CMake is re-run, MSBuild runs and notes everything is up to date, then the installation process is rerun. Linux will recognise everything is clean and do nothing. I suspect this is a difference in MSBuild vs. make/ninja CMake generators. That is, `Visual Studio CMake generator + MSBuild = more work done`.
